### ISSUE TITLE
Unable to compile MongoDB backend plugin

### ISSUE THREAD

**Issue by akashtalole** [jpmens/mosquitto-auth-plug#168]
mongoc.h: No such file or directory error occurs when compiling with mongodb.

**Comment by bitfreak25** [comment]
See #162 . This is solved in the master-branch of mosquitto-auth-plug.

**Comment by akashtalole** [comment]
Thanks!! Its working fine with new master code.

**Comment by akashtalole** [comment]
Load error: libmongoc-1.0.so.0: cannot open shared object file: No such file or directory

**Comment by akashtalole** [comment]
When I run like $ LD_LIBRARY_PATH=/usr/local/lib /usr/local/sbin/mosquitto -c /etc/mosquitto/mosquitto.conf . It works fine. How i can run it normally without  LD_LIBRARY_PATH

**Comment by bitfreak25** [comment]
I need to recompile the [mongo-c-driver](https://github.com/mongodb/mongo-c-driver) it on my own, cause there was no "libmongoc"-Package. After that I had to type

> ldconfig

 in the console, to finish it correctly.

**Comment by akashtalole** [comment]
@bitfreak25 working fine when type **ldconfig**.

**Comment by TheCherry** [comment]
i get the follow error, if I try to compile (installed mongo-c-driver 1.5.0):

`mosquitto-auth-plug # make
Selected backends:          MongoDB
Using mosquitto source dir: 
OpenSSL install dir:        /usr

If you changed the backend selection, you might need to 'make clean' first

CFLAGS:  -I/src/ -I/lib/ -fPIC -Wall -Werror  -DBE_MONGO  -I/usr/local/include/ -I/usr/local/include/libmongoc-1.0/ -I/usr/local/include/libbson-1.0/ -I/src -DDEBUG=1 -I/usr/include
LDFLAGS:  -L/usr/local/lib -L/lib/
LDADD:    -lmongoc-1.0 -lbson-1.0 -L/usr/lib -lcrypto -lmosquitto

cc -I/src/ -I/lib/ -fPIC -Wall -Werror  -DBE_MONGO  -I/usr/local/include/ -I/usr/local/include/libmongoc-1.0/ -I/usr/local/include/libbson-1.0/ -I/src -DDEBUG=1 -I/usr/include   -c -o be-mongo.o be-mongo.c
be-mongo.c: In function ‘be_mongo_getuser’:
be-mongo.c:141:4: error: ‘mongoc_collection_find’ is deprecated (declared at /usr/local/include/libmongoc-1.0/mongoc-collection.h:144): Use mongoc_collection_find_with_opts instead [-Werror=deprecated-declarations]
    cursor = mongoc_collection_find (collection,
    ^
be-mongo.c: In function ‘be_mongo_superuser’:
be-mongo.c:212:2: error: ‘mongoc_collection_find’ is deprecated (declared at /usr/local/include/libmongoc-1.0/mongoc-collection.h:144): Use mongoc_collection_find_with_opts instead [-Werror=deprecated-declarations]
  cursor = mongoc_collection_find(collection,
  ^
be-mongo.c: In function ‘be_mongo_aclcheck’:
be-mongo.c:262:2: error: ‘mongoc_collection_find’ is deprecated (declared at /usr/local/include/libmongoc-1.0/mongoc-collection.h:144): Use mongoc_collection_find_with_opts instead [-Werror=deprecated-declarations]
  cursor = mongoc_collection_find(collection,
  ^
be-mongo.c:286:5: error: ‘mongoc_collection_find’ is deprecated (declared at /usr/local/include/libmongoc-1.0/mongoc-collection.h:144): Use mongoc_collection_find_with_opts instead [-Werror=deprecated-declarations]
     cursor = mongoc_collection_find(collection,
     ^
cc1: all warnings being treated as errors
<builtin>: recipe for target 'be-mongo.o' failed
make: *** [be-mongo.o] Error 1
`

**Comment by hahasheminejad** [comment]
@TheCherry I had exactly the same problem and I could fix it by compiling mongo-auth using an older version of mongo-c-driver (1.2 worked for me). 
here is the sequences :

git clone https://github.com/mongodb/mongo-c-driver
cd mongo-c-driver/
make uninstall #if you already installed it
git checkout r1.2
./autogen.sh --with-libbson=bundled
make
make install
# after a successful install, you can proceed with compiling mosquitto-auth-plug but plz be inform to do make clean

**Comment by mouaadd** [comment]
i  get the follow error, if I try to compile (installed mongo-c-driver 1.6.0):

be-mongo.c: In function ‘be_mongo_getuser’:
be-mongo.c:141:4: error: ‘mongoc_collection_find’ is deprecated: Use mongoc_collection_find_with_opts instead [-Werror=deprecated-declarations]
    cursor = mongoc_collection_find (collection,
    ^
In file included from /usr/local/include/libmongoc-1.0/mongoc-client.h:27:0,
                 from /usr/local/include/libmongoc-1.0/mongoc.h:27,
                 from be-mongo.c:14:
/usr/local/include/libmongoc-1.0/mongoc-collection.h:143:1: note: declared here
 mongoc_collection_find (mongoc_collection_t *collection,
 ^
be-mongo.c: In function ‘be_mongo_superuser’:
be-mongo.c:212:2: error: ‘mongoc_collection_find’ is deprecated: Use mongoc_collection_find_with_opts instead [-Werror=deprecated-declarations]
  cursor = mongoc_collection_find(collection,
  ^
In file included from /usr/local/include/libmongoc-1.0/mongoc-client.h:27:0,
                 from /usr/local/include/libmongoc-1.0/mongoc.h:27,
                 from be-mongo.c:14:
/usr/local/include/libmongoc-1.0/mongoc-collection.h:143:1: note: declared here
 mongoc_collection_find (mongoc_collection_t *collection,
 ^
be-mongo.c: In function ‘be_mongo_aclcheck’:
be-mongo.c:262:2: error: ‘mongoc_collection_find’ is deprecated: Use mongoc_collection_find_with_opts instead [-Werror=deprecated-declarations]
  cursor = mongoc_collection_find(collection,
  ^
In file included from /usr/local/include/libmongoc-1.0/mongoc-client.h:27:0,
                 from /usr/local/include/libmongoc-1.0/mongoc.h:27,
                 from be-mongo.c:14:
/usr/local/include/libmongoc-1.0/mongoc-collection.h:143:1: note: declared here
 mongoc_collection_find (mongoc_collection_t *collection,
 ^
be-mongo.c:286:5: error: ‘mongoc_collection_find’ is deprecated: Use mongoc_collection_find_with_opts instead [-Werror=deprecated-declarations]
     cursor = mongoc_collection_find(collection,
     ^
In file included from /usr/local/include/libmongoc-1.0/mongoc-client.h:27:0,
                 from /usr/local/include/libmongoc-1.0/mongoc.h:27,
                 from be-mongo.c:14:
/usr/local/include/libmongoc-1.0/mongoc-collection.h:143:1: note: declared here
 mongoc_collection_find (mongoc_collection_t *collection,
 ^
cc1: all warnings being treated as errors
<builtin>: recipe for target 'be-mongo.o' failed
make: *** [be-mongo.o] Error 1

**Comment by openbufo** [comment]
In case you are still facing issue with compiling the plugin with MongoDB backend, make sure you are using 1.6 version of mongo c driver. With commit #8ed546cf567c1fc02e4386310949537d0781592b, depereciated mongo calls have been replaced. So, you need to have the latest version of the driver.

```
git clone https://github.com/mongodb/mongo-c-driver
cd mongo-c-driver/
git checkout r1.6
cd mongo-c-driver
./autogen.sh --with-libbson=bundled
make
make install
make clean
ldconfig
```
Hope this resolves the issue.
